timers: do not expose .unref()._handle._list

Fishrock123 opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->
##### Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] `make -j4 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines
##### Affected core subsystem(s)

<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

timers
##### Description of change

<!-- Provide a description of the change below this comment. -->

The list is defunct at this point.
I believe this to be in effect a minor defect/oversight from
3eecdf9f1422640c7439507cb39ffb3dff57b3e4

Tangentially effects/related to https://github.com/nodejs/node/pull/8372

Edit: cc @indutny?

YurySolovyov commented on lib/timers.js:
```diff
@@ -481,6 +481,7 @@ Timeout.prototype.unref = function() {
     }
 
     var handle = reuse(this);
+    delete handle._list;
```
I thought V8 don't like deleting props prom objects, but it might be ok if it is not perf-sensitive...

Fishrock123 replied:
Under https://github.com/nodejs/node/pull/8372 this code path is far less likely to be hit in the first place.

I doubt the overhead from a later-than-this-tick `unref()` is much of a worry.

misterdjules replied:
Why not removing the reference to the `TimersList` instance in `reuse`'s implementation instead? 

Also, #8372 hasn't landed yet, and it's not clear whether it will land. As a result I would think that if deleting the `._list` property is more costly than setting it to `null` (or doing anything else that we know might perform better), we should not do that and instead do what we know performs better.

Fishrock123 replied:
> Why not removing the reference to the `TimersList` instance in `reuse`'s implementation instead?

Messes with some tests that check thing after `unenroll()`, which also uses `reuse()`.

--

Setting it to `undefined` would be faster but I figured it could be confusing that sometimes there would be a `_list` property.

If that is acceptable we can switch to it though.

misterdjules replied:
> > Why not removing the reference to the TimersList instance in reuse's implementation instead?

> Messes with some tests that check things after unenroll(), which also uses reuse().

Making that change (clearing the handle's reference to the `TimersList` instance) makes only one test fail on `master`: `test/parallel/test-timers-same-timeout-wrong-list-deleted.js`. I think that test could be rewritten to not rely on that reference being present for unenrolled timers.

Is this what you had in mind by "Messes with some tests"?

Fishrock123 replied:
> Is this what you had in mind by "Messes with some tests"?

Yeah, besides the fact that it is far more common and can more easily induce a perf problem.

misterdjules replied:
> Yeah, besides the fact that it is far more common and can more easily induce a perf problem

Have we measured that potential performance problem?

Fishrock123 replied:
`delete` is known to be a heavy operation.

misterdjules replied:
@Fishrock123 

> delete is known to be a heavy operation.

I thought we were also considering setting the property to `null` or `undefined`, isn't it still the case?

You mentioned earlier:

> Setting it to undefined would be faster but I figured it could be confusing that sometimes there would be a _list property.

> If that is acceptable we can switch to it though.

Having a `_list` property set to `null` or `undefined` seems like it would be acceptable.

indutny replied:
Should we just use `Symbol`s for `_list`?

Fishrock123 replied:
We could, I thought those were exposable somehow too?

I still think we should unreference it in some way though so the list can be GC'd.

jasnell replied:
Using a Symbol *and* setting the value to undefined would accomplish both goals without incurring the potential perf hit of the delete, yes? Symbols are exposable but not without extra work.

trevnorris replied:
or, why don't we set it to `null`. we already do that for some other properties like `_handle`.

Fishrock123 commented:
cc also @misterdjules maybe?

misterdjules commented:
The idea seems reasonable, I added a couple comments regarding the implementation.

misterdjules commented:
It seems that removing the `._list` property would be a breaking change. Setting it to `null` or `undefined` could be considered as a breaking change too. As a result it seems this change should be labelled `semver-major`.

Fishrock123 commented:
> It seems that removing the `._list` property would be a breaking change. 

Theoretically but I'm willing to bet no one even knew it existed.

misterdjules commented:
@Fishrock123 

> > It seems that removing the ._list property would be a breaking change.

> Theoretically but I'm willing to bet no one even knew it existed.

Can we use https://github.com/ChALkeR/Gzemnid to look into that? I would like to err on the side of caution regardless of the results, unless the benefits of the change far outweigh the risks, however unlikely.

Fishrock123 commented:
cc @ChALkeR?

I dunno how this would be detected, but I _guess_ you could look for `._list`?

ChALkeR commented:
There are a lot of stuff on '\._list[^a-zA-Z0-9_]', but the ones I seen don't look timers related.
Also a lot of stuff on '\._handle[^a-zA-Z0-9_]'.

Nothing on `handle\._list`, but that doesn't guarantee anything.
I will take a closer look soon.

Fishrock123 commented:
@nodejs/ctc Anyone object to me merging this one as a `major`?

The `_handle._list` double-private property is only exposed under a specific condition and is non-functional when it is exposed.

targos commented:
+1

trevnorris commented:
@Fishrock123 If it's non-functional when it's exposed then I'd consider this a bug fix. Or at most a minor. Calling this a major seems like far overkill. We're removing a property that's _broken_ right?

Fishrock123 commented:
@trevnorris The property exists, but it doesn't actually have any relevant functionality since it is circularly attached to a handle which is repurposed for an individual timeout.

Fishrock123 commented:
@indutny What's the point? This is the only spot where it is exposed to a user.

jasnell commented:
ping @Fishrock123 ... still making progress on this?

jasnell commented:
Ping @Fishrock123 ...

Fishrock123 commented:
@indutny Is there a large benefit here to using Symbols rather than just deleting it?

I mean with any property symbol or not I guess we could just set it to `undefined` but it'd be weird that it is there at all imo. Could work though...

Fishrock123 commented:
Updated to null out the property. New CI: https://ci.nodejs.org/job/node-test-pull-request/7020/

@indutny, @misterdjules lgty?

@nodejs/ctc please take a quick look if you could since it is semver-major.

Fishrock123 commented:
Oops, actually new CI: https://ci.nodejs.org/job/node-test-pull-request/7021/

misterdjules commented on lib/timers.js:
```diff
@@ -538,6 +538,7 @@ Timeout.prototype.unref = function() {
     }
 
     var handle = reuse(this);
+    handle._list = null;
```
What happens when `handle` is `null`?

Fishrock123 commented:
Updated, ptal @misterdjules, @indutny 

New CI: https://ci.nodejs.org/job/node-test-pull-request/7303/

the test does demonstrate now why I think it makes more sense as `undefined`, though. I'll change that if we can agree.

misterdjules commented on lib/timers.js:
```diff
@@ -538,6 +538,7 @@ Timeout.prototype.unref = function() {
     }
 
     var handle = reuse(this);
+    if (handle) handle._list = null;
```
nit: I find the:
```
if (handle) {
  handle._list = null;
}
```
form clearer and less error-prone.

Fishrock123 replied:
I agree, but we don't seem to care about it in the rest of the codebase...
¯\\\_(ツ)\_/¯

misterdjules replied:
Sounds good, that's just a nit, so I'm fine to keep it as it is currently.

misterdjules commented:
@Fishrock123 As you suggested, I'd prefer the `_list` property to have the same value for unrefed timers, regardless of whether the underlying `TimerWrap` instance was reused.

Fishrock123 commented:
Updated, CI again: https://ci.nodejs.org/job/node-test-pull-request/7316/

misterdjules commented on test/parallel/test-timers-unref-reuse-no-exposed-list.js:
```diff
@@ -0,0 +1,16 @@
+'use strict';
```
Does this need a license block now that https://github.com/nodejs/node/pull/10155 landed?

Fishrock123 replied:
... No? It did not have previous attribution? (It's new?)

misterdjules replied:
OK, it's been a while I haven't added a file to the repository, and it seems files added in the recent past did not include a license block, so it sounds good to me!

misterdjules commented on test/parallel/test-timers-unref-reuse-no-exposed-list.js:
```diff
@@ -0,0 +1,16 @@
+// Check that everything works even if the handle was not re-used.
+setTimeout(() => {}, 1);
+const timer2 = setTimeout(() => {}, 1).unref();
+// Note: It is unlikely that this assertion will be hit in a failure.
```
That comment seems to be specific to a previous version of this PR where it would throw in that case. I don't think it's relevant in the final version.

Fishrock123 replied:
Not really, the comment states if that were to fail, i.e. the conditional was removed, node will explode internally rather than triggering this assertion.

misterdjules replied:
Why would "a failure" be _specifically_ that "the conditional was removed"? This is anticipating the future, and not reflecting the current state of the code base. Without the context of what this PR looked like in an intermediate version, this comment doesn't make any sense.

I still think it's not relevant.

Fishrock123 replied:
I can't think of how this will change without removing handle re-use... Are you saying that a comment that either: might be useful or not matter at all would be better off not being there at all? Isn't that every comment ever?

That bit of code does primarily test that conditional though, that isn't a problem IMO. We have added plenty of test code for code coverage that would be similar.

misterdjules replied:
The test code itself is great, and the comment above:

> // Check that everything works even if the handle was not re-used.

is both necessary and sufficient to explain why this test code is present.

On the other hand,  I find the comment:

> // Note: It is unlikely that this assertion will be hit in a failure.
> // Instead, timers.js will likely throw a TypeError internally.

confusing and detrimental to the understanding of that test because it does not reflect the current state of the implementation. Instead it describes what could happen if the code was changed in a specific way, that is if a conditional was removed.

However, I don't think we want to start documenting every possible way this test could fail depending on an infinite list of possible changes, or by definition we will end up with an infinite list of comments.

Fishrock123 commented:
@misterdjules I believe I have addressed your comments.

@indutny last chance to look, I assume you are not exactly interested so I'll be dismissing your review as it was long since addressed.

Fishrock123 commented:
@misterdjules updated, lgty?

Fishrock123 commented:
Here's a CITGM run although I'd be far beyond surprised if we covered a module interacting with this in any way. https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/719/

BridgeAR commented:
I dismissed the review of @indutny. @nodejs/tsc PTAL because of that. This should be ready to land otherwise.

mcollina commented:
let’s land this.

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10153/

BridgeAR commented:
Landed in bbbfbcbb88148f24e9db8ee7828dd467e3335ad4
